src/main/java/bio/terra/common/kubernetes/KubeService.java

```diff
@@ -0,0 +1,190 @@
+  static final String KUBE_NAMESPACE_FILE = KUBE_DIR + "/namespace";
+
+  private final String podName;
+  private final boolean inKubernetes;
+  private final String podNameFilter;
+  private final String namespace;
+  private final AtomicBoolean isShutdown;
+
+  private KubePodListener podListener;
+  private Thread podListenerThread;
+
+  /**
+   * @param podName Name of the Kubernetes pod we are running in. If we are not in a pod, this
+   *     defaults to a constant string in application properties. We cannot easily find our own pod
+   *     name, so components must plumb this in using settings in the helm charts.
+   * @param inKubernetes Used to denote that we are running in the Kubernetes environment. The
+   *     service can be called and provides reasonable answers when inKubernetes is false.
+   * @param podNameFilter Filter to apply to pod names returned by Kubernetes API to know which pods
+   *     we care about.
+   */
+  public KubeService(String podName, boolean inKubernetes, String podNameFilter) {
+    this.podName = podName;
+    this.inKubernetes = inKubernetes;
+    this.podNameFilter = podNameFilter;
+    if (inKubernetes) {
+      this.namespace = readFileIntoString(KUBE_NAMESPACE_FILE);
+    } else {
+      this.namespace = "nonamespace";
+    }
+    this.isShutdown = new AtomicBoolean(false);
+
+    logger.info(
+        "Kubernetes configuration: inKube: "
+            + inKubernetes
+            + "; namespace: "
+            + namespace
+            + "; podName: "
+            + podName);
+  }
+
+  /**
+   * Get a list of the API pods from Kubernetes. It is returned as a set to make probing for
+   * specific names easy.
+   *
+   * @return set of pod names containing the podNameFilter string; null if not in kubernetes
+   */
+  public Set<String> getPodList() {
+    Set<String> pods = new HashSet<>();
+    if (!inKubernetes) {
+      return pods;
+    }
+
+    try {
+      CoreV1Api api = makeCoreApi();
+      V1PodList list =
+          api.listNamespacedPod(namespace, null, null, null, null, null, null, null, null, null);
+      for (V1Pod item : list.getItems()) {
+        if (item.getMetadata() != null) {
+          String podName = item.getMetadata().getName();
+          if (StringUtils.contains(podName, podNameFilter)) {
+            pods.add(podName);
+          }
+        }
+      }
+      return pods;
+    } catch (ApiException ex) {
+      throw new KubeApiException("Error listing pods", ex);
+    }
+  }
+
+  /**
+   * Launch the {@link KubePodListener} thread.
+   *
+   * @param stairway the Stairway instance passed to the pod listener for Stairway recovery
+   */
+  public void startPodListener(Stairway stairway) {
+    if (inKubernetes) {
+      podListener = new KubePodListener(this, stairway, namespace, podNameFilter);
+      podListenerThread = new Thread(podListener);
+      podListenerThread.start();
+    }
+  }
+
+  public int getActivePodCount() {
+    if (podListener != null) {
```

Comment:
I'd use `Optional.ofNullable().map().orElse()` here.

Reply:
I would never use that. Why is everyone so twitchy about comparing to null? What is the benefit of running through 4 methods to do the same thing?

Reply:
It's more deliberate, and a higher level of abstraction. There's no need to branch, and you're very clear about what you're getting and that it might not be there and what the default is.

`if` is basically deprecated as of Java 8.